macOS: Adjust segment offset in the core file

knn-k opened the pull request:
This commit adjusts the offsets of segments in the core file when the
segment is unreadable.

Signed-off-by: KONNO Kazuhiro <konno@jp.ibm.com>

knn-k commented:
This PR resolves https://github.com/eclipse-openj9/openj9/issues/14871

knn-k commented:
jenkins build osx

0xdaryl commented:
@keithc-ca or @babsingh : would you mind reviewing this please?

knn-k commented:
See an example of an unreadable segment in a core file for AArch64 macOS after this change is applied:
```
Load command 37
      cmd LC_SEGMENT_64
  cmdsize 72
  segname
   vmaddr 0x0000000100240000
   vmsize 0x0000000000004000
  fileoff 285544
 filesize 0
  maxprot 0x00000007
 initprot 0x00000000
   nsects 0
    flags 0x0
Load command 38
      cmd LC_SEGMENT_64
  cmdsize 72
  segname
   vmaddr 0x0000000100244000
   vmsize 0x0000000000008000
  fileoff 285544
 filesize 32768
  maxprot 0x00000007
 initprot 0x00000003
   nsects 0
    flags 0x0
```
The segment of `Load command 37` is unreadable, and its `filesize` field is set to zero.  `Load command 38` has the same `fileoff` value as that of `Load command 37`.

knn-k commented:
I ran sanity.functional with OpenJ9 for x86_64 macOS in an internal job, to make sure this change does not break core files on x86_64.  The test job finished successfully.
(internal job/Test_openjdk17_j9_sanity.functional_x86-64_mac_Personal/122)

pshipton commented:
It would be good to run extended.functional as well.

knn-k commented:
I started an extended.functional job for x86_64 macOS: job/Grinder/22475.

keithc-ca commented on port/osx/omrosdump.c:
```diff
@@ -210,25 +211,27 @@ coredump_to_file(mach_port_t task_port, pid_t pid)
 		}
 
 		segments[i].fileoff = seg_file_off;
+		segments[i].filesize = bytes_written;
```
I believe `bytes_written == bytes_read` is true here and so we don't need to introduce `bytes_written` and can instead simply use `bytes_read` here and below.

knn-k replied:
I stopped introducing `bytes_written` as suggested.

knn-k commented:
I see many test failures in extended.functional on x86 macOS.  All of them are UnsatisfiedLinkError for libawt.dylib.
Here is an example from cmdLineTester_classesdbgddrext_1:
```
 [OUT] DTFJView version 4.29.5, using DTFJ version 1.12.29003
 [OUT] Loading image from DTFJ...
 [OUT]
 [ERR] Exception in thread "main" java.lang.UnsatisfiedLinkError: Failed to load library "/Users/jenkins/workspace/Grinder_testList_1/openjdkbinary/j2sdk-image/lib/libawt.dylib"
 [ERR] 	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
 [ERR] 	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:388)
 [ERR] 	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:232)
 [ERR] 	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:174)
 [ERR] 	at java.base/jdk.internal.loader.NativeLibraries.findFromPaths(NativeLibraries.java:315)
 [ERR] 	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:285)
 [ERR] 	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:1777)
 [ERR] 	at java.base/java.lang.System.loadLibrary(System.java:729)
```

pshipton commented:
What version of the jdk, and what OS version?
FYI https://github.com/eclipse-openj9/openj9-docs/pull/912
and infrastructure/issues/5416

knn-k commented:
It was an internal Grinder job job/Grinder/22475 on mac11x64rt01, which seems to run macOS 11.3 for x86_64.
JDK version is 17.
```
[2022-04-08T12:59:13.555Z] openjdk version "17.0.3-internal" 2022-04-19
[2022-04-08T12:59:13.555Z] OpenJDK Runtime Environment (build 17.0.3-internal+0-adhoc.jenkins.BuildJDK17x86-64macPersonal)
[2022-04-08T12:59:13.555Z] Eclipse OpenJ9 VM (build master-8a7ca229b6b, JRE 17 Mac OS X amd64-64-Bit Compressed References 20220408_146 (JIT enabled, AOT enabled)
[2022-04-08T12:59:13.555Z] OpenJ9   - 8a7ca229b6b
[2022-04-08T12:59:13.555Z] OMR      - 99c7cc228d5
[2022-04-08T12:59:13.555Z] JCL      - aff77b6333b based on jdk-17.0.3+5)
```

knn-k commented on port/osx/omrosdump.c:
```diff
@@ -210,25 +211,27 @@ coredump_to_file(mach_port_t task_port, pid_t pid)
 			}
+
+			written = pwrite(corefile_fd, (void *)data_read, data_size, seg_file_off + bytes_read);
 			if (written < 0) {
```
I updated the code as suggested.

knn-k commented:
I updated the code, and now it passes extended.functional on x86_64 macOS (job/Test_openjdk17_j9_extended.functional_x86-64_mac_Personal/112).
I stopped modifying `segments[i].filesize`.
I also updated the commit title and the commit message: "segment size" -> "segment offset".

keithc-ca commented on port/osx/omrosdump.c:
```diff
@@ -210,10 +210,10 @@ coredump_to_file(mach_port_t task_port, pid_t pid)
 			}
 			if (to_read != data_size) {
 				break;
```
If a segment is not read in its entirety (e.g. via the `break` here), we need
```
    segments[i].filesize = bytes_read;
```
after line 223 (as was present as a comment in ef5cdbaa01c7628a4d96ade836f1565d37906d80).

knn-k replied:
I originally thought I needed the line `segments[i].filesize = bytes_read;` for the case `bytes_read < segments[i].filesize`.
But I removed the line because it caused UnsatisfiedLinkErrors in an x86_64 test job as I described above.

keithc-ca replied:
Without it, there's the possibility that `segments[i].filesize` is too large, and the segment will extend into the file space for following segments (or beyond the end of the file).

keithc-ca replied:
Perhaps the `UnsatisfiedLinkError` will be resolved by adoptium/aqa-tests#3559?

pshipton replied:
Only if running JCKs. We'd have to add the option to the functionality test(s). Plus that is a jdk8 problem afaik.

We ran a number of tests for this specific issue, see runtimes/backlog/issues/742#issuecomment-42417615, but they didn't run on mac11x64rt01. It's possible there is a machine problem. We have had missing prereqs in the past on some platforms runtimes/backlog/issues/742#issuecomment-42407773. There is a list of the prereqs required for jdk18 Mac at runtimes/infrastructure/issues/5416#issuecomment-42508170.

knn-k replied:
I added `segments[i].filesize = bytes_read;` back again.

knn-k replied:
I ran extended.functional on mac10x64rt03 and on osxrt2, and got no `UnsatisfiedLinkError`.

knn-k commented:
I ran extended.functional on x86_64 macOS (job/Test_openjdk11_j9_extended.functional_x86-64_mac_Personal/133).  There is no failures related to DDR in it.  One failure with rc021 in cmdLineTester_jvmtitests_hcr_OSRG_nongold_0 looks like a known problem https://github.com/eclipse-openj9/openj9/issues/14115.

knn-k commented:
One of unreadable segments and the last segment in `otool` output for a core file on x86_64 macOS look like this, with this PR applied:
```
Load command 120
      cmd LC_SEGMENT_64
  cmdsize 72
  segname
   vmaddr 0x00000000116f6000
   vmsize 0x0000000004800000
  fileoff 128873720
 filesize 0
  maxprot 0x00000007
 initprot 0x00000000
   nsects 0
    flags 0x0
```

```
Load command 266
      cmd LC_SEGMENT_64
  cmdsize 72
  segname
   vmaddr 0x00007ffffff69000
   vmsize 0x0000000000001000
  fileoff 3373815032
 filesize 4096
  maxprot 0x00000005
 initprot 0x00000005
   nsects 0
    flags 0x0
```

The size of the core file is 3,373,819,128 bytes, which is the same as `fileoff` + `filesize` in the last segment.
There are no "past end of file" messages in the `otool` output.

0xdaryl commented:
Jenkins build osx
